File: python/LongestCommonPrefix.py
def longest_common_prefix(arr):
    size = float("inf")

    # find the shortest string length, since lcp cannot be longer than this
    for i in arr:
        size = min(size, len(i))

    # make pre to be first string in array
    pre = arr[0]

    # outer: loop through arr[1] to end
    for i in xrange(1, len(arr)):
        tmp = arr[i]

        # loop through all chars in array up to size
        for j in xrange(size):
            # if char j cannot be find in pre, cut pre up to j
            if pre.find(tmp[j]) == -1:
                pre = pre[:j]

    print pre


if __name__ == '__main__':
    arr = ["aaa", "aab", "aac", "aaad"]
    longest_common_prefix(arr)

File: java/LongestCommonPrefix.java
/**
 * Created by adamli on 8/8/16.
 */
public class LongestCommonPrefix {
    String LongestCommonPrefix(String[] list) {
        int len = Integer.MAX_VALUE;

        // get the shortest string length
        for (String str : list) {
            if (str.length() < len)
                len = str.length();
        }

        // set longest common prefix to be first string
        String rst = list[0];

        // outer loop: loop through words
        for (int i = 1; i < list.length; i++) {
            // inner loop : loop through chars in word
            for (int j = 0; j < len; j++) {
                if (list[i].charAt(j) != rst.charAt(j)) {
                    rst = rst.substring(0, j);
                }
            }
        }

        return rst;
    }
}
